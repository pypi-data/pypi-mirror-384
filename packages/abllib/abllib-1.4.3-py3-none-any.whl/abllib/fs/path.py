"""A module containing path-modification functions."""

import pathlib

def absolute(*paths: str | pathlib.Path) -> str:
    """
    Return an absolute path, regardless of what is input.

    Additionally, the path is resolved, removing any symlinks on the way.
    """

    if len(paths) == 0:
        raise ValueError()
    for item in paths:
        if not isinstance(item, (str, pathlib.Path)):
            raise TypeError()

    path = pathlib.Path(*paths)

    if path.is_absolute():
        return str(path.resolve())

    return str(path.absolute().resolve())
