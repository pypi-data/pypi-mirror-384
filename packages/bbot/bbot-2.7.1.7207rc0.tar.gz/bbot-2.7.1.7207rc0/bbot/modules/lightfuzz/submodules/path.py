from .base import BaseLightfuzz
from bbot.errors import HttpCompareError

from urllib.parse import quote


class path(BaseLightfuzz):
    """
    Detects path traversal and local file inclusion vulnerabilities

    Techniques:

    * Relative Path Traversal:
       - Tests various relative path traversal patterns (../, ./, .../, etc.)
       - Uses multiple encoding variations (URL encoding, double encoding)
       - Attempts various path validation bypass techniques

    * Absolute Path Traversal:
       - Tests absolute paths for Windows (c:\\windows\\win.ini)
       - Tests absolute paths for Unix (/etc/passwd)
       - Tests null byte injection for extension bypass (%00)

    Results are validated using multiple confirmations and WAF response filtering to eliminate false positives.
    """

    friendly_name = "Path Traversal"

    async def fuzz(self):
        cookies = self.event.data.get("assigned_cookies", {})
        probe_value = self.incoming_probe_value(populate_empty=False)
        if not probe_value:
            self.debug(
                f"Path Traversal detection requires original value, aborting [{self.event.data['type']}] [{self.event.data['name']}]"
            )
            return

        # Single dot traversal tolerance test
        path_techniques = {
            "single-dot traversal tolerance (no-encoding)": {
                "singledot_payload": f"./a/../{probe_value}",
                "doubledot_payload": f"../a/../{probe_value}",
            },
            "single-dot traversal tolerance (no-encoding, leading slash)": {
                "singledot_payload": f"/./a/../{probe_value}",
                "doubledot_payload": f"/../a/../{probe_value}",
            },
            "single-dot traversal tolerance (url-encoding)": {
                "singledot_payload": quote(f"./a/../{probe_value}".encode(), safe=""),
                "doubledot_payload": quote(f"../a/../{probe_value}".encode(), safe=""),
            },
            "single-dot traversal tolerance (url-encoding, leading slash)": {
                "singledot_payload": quote(f"/./a/../{probe_value}".encode(), safe=""),
                "doubledot_payload": quote(f"/../a/../{probe_value}".encode(), safe=""),
            },
            "single-dot traversal tolerance (non-recursive stripping)": {
                "singledot_payload": f"...//a/....//{probe_value}",
                "doubledot_payload": f"....//a/....//{probe_value}",
            },
            "single-dot traversal tolerance (non-recursive stripping, leading slash)": {
                "singledot_payload": f"/...//a/....//{probe_value}",
                "doubledot_payload": f"/....//a/....//{probe_value}",
            },
            "single-dot traversal tolerance (double url-encoding)": {
                "singledot_payload": f".%252fa%252f..%252f{probe_value}",
                "doubledot_payload": f"..%252fa%252f..%252f{probe_value}",
            },
            "single-dot traversal tolerance (double url-encoding, leading slash)": {
                "singledot_payload": f"%252f.%252fa%252f..%252f{probe_value}",
                "doubledot_payload": f"%252f..%252fa%252f..%252f{probe_value}",
            },
        }

        compiled_regex = self.lightfuzz.helpers.re.compile(r"/(?:[\w-]+/)*[\w-]+\.\w+")
        linux_path_regex = await self.lightfuzz.helpers.re.match(compiled_regex, probe_value)
        if linux_path_regex is not None:
            original_path_only = "/".join(probe_value.split("/")[:-1])
            original_filename_only = probe_value.split("/")[-1]
            # Some servers validate the start of the path, so we construct our payload with the original path and filename
            path_techniques["single-dot traversal tolerance (start of path validation)"] = {
                "singledot_payload": f"{original_path_only}/./{original_filename_only}",
                "doubledot_payload": f"{original_path_only}/../{original_filename_only}",
            }

        for path_technique, payloads in path_techniques.items():
            iterations = 5  # one failed detection is tolerated, as long as its not the first run
            confirmations = 0
            while iterations > 0:
                try:
                    http_compare = self.compare_baseline(
                        self.event.data["type"], probe_value, cookies, skip_urlencoding=True
                    )
                    singledot_probe = await self.compare_probe(
                        http_compare,
                        self.event.data["type"],
                        payloads["singledot_payload"],
                        cookies,
                        skip_urlencoding=True,
                    )
                    doubledot_probe = await self.compare_probe(
                        http_compare,
                        self.event.data["type"],
                        payloads["doubledot_payload"],
                        cookies,
                        skip_urlencoding=True,
                    )
                    # if singledot_probe[0] is true, the response is the same as the baseline. This indicates adding a single dot did not break the functionality
                    # next, if doubledot_probe[0] is false, the response is different from the baseline. This further indicates that a real path is being manipulated
                    # if doubledot_probe[3] is not None, the response is not empty.
                    # if doubledot_probe[1] is not ["header"], the response is not JUST a header change.
                    # "The requested URL was rejected" is a very common WAF error message which appears on 200 OK response, confusing detections
                    if (
                        singledot_probe[0] is True
                        and doubledot_probe[0] is False
                        and doubledot_probe[3] is not None
                        and doubledot_probe[1] != ["header"]
                        and "The requested URL was rejected" not in doubledot_probe[3].text
                    ):
                        confirmations += 1
                        self.verbose(f"Got possible Path Traversal detection: [{str(confirmations)}] Confirmations")
                        # only report if we have 3 confirmations
                        if confirmations > 3:
                            self.results.append(
                                {
                                    "type": "FINDING",
                                    "description": f"POSSIBLE Path Traversal. {self.metadata()} Detection Method: [{path_technique}]",
                                }
                            )
                            # no need to report both techniques if they both work
                            break
                except HttpCompareError as e:
                    iterations -= 1
                    self.debug(e)
                    continue

                iterations -= 1
                if confirmations == 0:
                    break

        # Absolute path test, covering Windows and Linux
        absolute_paths = {
            r"c:\\windows\\win.ini": "; for 16-bit app support",
            "/etc/passwd": "daemon:x:",
            "../../../../../etc/passwd%00.png": "daemon:x:",
        }

        for path, trigger in absolute_paths.items():
            r = await self.standard_probe(self.event.data["type"], cookies, path, skip_urlencoding=True)
            if r and trigger in r.text:
                self.results.append(
                    {
                        "type": "FINDING",
                        "description": f"POSSIBLE Path Traversal. {self.metadata()} Detection Method: [Absolute Path: {path}]",
                    }
                )
