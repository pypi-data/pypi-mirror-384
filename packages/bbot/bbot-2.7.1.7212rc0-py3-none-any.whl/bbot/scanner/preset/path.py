import logging
from pathlib import Path

from bbot.errors import *

log = logging.getLogger("bbot.presets.path")

DEFAULT_PRESET_PATH = Path(__file__).parent.parent.parent / "presets"
DEFAULT_PRESET_PATH = DEFAULT_PRESET_PATH.expanduser().resolve()


class PresetPath:
    """
    Keeps track of where to look for preset .yaml files
    """

    def __init__(self):
        self.paths = [DEFAULT_PRESET_PATH]

    def find(self, filename):
        filename_path = Path(filename).expanduser()
        extension = filename_path.suffix.lower()
        file_candidates = set()
        extension_candidates = {".yaml", ".yml"}
        if extension:
            extension_candidates.add(extension.lower())
        else:
            file_candidates.add(filename_path.stem)
        for ext in extension_candidates:
            file_candidates.add(f"{filename_path.stem}{ext}")
        file_candidates = sorted(file_candidates)
        file_candidates_str = ",".join([str(s) for s in file_candidates])
        if "/" in str(filename):
            self.add_path(filename_path.parent)
        log.debug(f"Searching for {file_candidates_str} in {[str(p) for p in self.paths]}")
        for path in self.paths:
            for candidate in file_candidates:
                for file in path.rglob(f"**/{candidate}"):
                    if file.is_file():
                        log.verbose(f'Found preset matching "{filename}" at {file}')
                        self.add_path(file.parent)
                        return file
        raise ValidationError(
            f'Could not find preset at "{filename}" - file does not exist. Use -lp to list available presets'
        )

    def __str__(self):
        return ":".join([str(s) for s in self.paths])

    def add_path(self, path):
        path = Path(path).expanduser().resolve()
        # skip if already in paths
        if path in self.paths:
            return
        # skip if path is a subdirectory of any path in paths
        if any(path.is_relative_to(p) for p in self.paths):
            return
        # skip if path is not a directory
        if not path.is_dir():
            log.debug(f'Path "{path.resolve()}" is not a directory')
            return
        # preemptively remove any paths that are subdirectories of the new path
        self.paths = [p for p in self.paths if not p.is_relative_to(path)]
        self.paths.insert(0, path)

    def __iter__(self):
        yield from self.paths


PRESET_PATH = PresetPath()
