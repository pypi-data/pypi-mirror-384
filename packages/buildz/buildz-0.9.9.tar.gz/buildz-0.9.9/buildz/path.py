from . import Base
import os,sys,json,time
dirname = os.path.dirname
join = os.path.join
class Key:
    first='first'
    ifirst = 0
    last= 'last'
    ilast = -1
    @staticmethod
    def check_first(val):
        return val in (Key.first, Key.ifirst)
    @staticmethod
    def check_last(val):
        return val in (Key.last, Key.ilast)
class Path(Base):
    Key = Key
    def init(self, *args, **maps):
        self.paths = {}
        self.fcs = {}
        self.lasts = {}
        k = "check_abs"
        self.check_abs = True
        if len(args)>0:
            self.set(args)
        if k in maps:
            self.check_abs = maps[k]
            del maps[k]
        for k, v in maps.items():
            self.set(v, k)
    @staticmethod
    def dir(fp):
        return os.path.dirname(fp)
    @staticmethod
    def join(*a):
        return os.path.join(*a)
    @staticmethod
    def rjoin(path, *a):
        a = [k.strip() for k in a if k is not None and k.strip()!=""]
        if len(a)==0:
            return path
        if path is None:
            return os.path.join(*a)
        return os.path.join(path, *a)
    @staticmethod
    def rfp(paths, *a, last=-1, check_abs=False):
        a = [k.strip() for k in a if k is not None and k.strip()!=""]
        # if len(a)==0:
        #     a = ""
        if check_abs and len(a)>0:
            f = a[0]
            if f[:1]=="/" or f.find(":")>0:
                return Path.join(*a)
        #print(f"[TESTZ] a: {a}, paths: {paths}")
        #fp = os.path.join(*a)
        for path in paths:
            _fp = Path.rjoin(path, *a)
            if os.path.exists(_fp):
                return _fp
        return Path.rjoin(paths[last], *a)
    def add(self, path, name=None, index=0):
        self.paths[name].insert(index, path)
    def set(self, paths, name=None, last = -1, curr = None):
        if type(paths) not in (list, tuple):
            paths = [paths]
        if curr is not None and None not in paths:
            if Key.check_first(curr):
                paths = [None]+paths
            elif Key.check_last(curr):
                paths.append(None)
            else:
                assert False
        self.paths[name] = paths
        self.lasts[name] = last
        def fc(*a):
            return self.rfp(paths, *a, last=last, check_abs=self.check_abs)
        self.fcs[name] = fc
    def __getattr__(self, name):
        return self.fcs[name]
    def call(self, *obj):
        return self.fcs[None](*obj)

pass
