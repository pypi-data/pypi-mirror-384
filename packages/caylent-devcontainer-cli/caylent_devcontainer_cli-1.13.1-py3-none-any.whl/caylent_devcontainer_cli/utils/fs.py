"""File system utilities for the Caylent Devcontainer CLI."""

import json
import os
from typing import Any, Dict, List

from caylent_devcontainer_cli.utils.ui import confirm_action, log


def load_json_config(file_path: str) -> Dict[str, Any]:
    """Load JSON configuration file."""
    try:
        with open(file_path, "r") as f:
            data = json.load(f)
        return data
    except Exception as e:
        log("ERR", f"Error loading {file_path}: {e}")
        import sys

        sys.exit(1)


def generate_exports(env_dict: Dict[str, Any], export_prefix: bool = True) -> List[str]:
    """Generate shell export statements from a dictionary."""
    lines = []
    for key, value in env_dict.items():
        if isinstance(value, (dict, list)):
            val = json.dumps(value)
        else:
            val = str(value)
        prefix = "export " if export_prefix else ""
        line = f"{prefix}{key}='{val}'"
        lines.append(line)
    return lines


def generate_shell_env(json_file: str, output_file: str, no_export: bool = False) -> None:
    """Generate shell environment file from JSON config."""
    log("INFO", f"Reading configuration from {json_file}")
    data = load_json_config(json_file)

    # Only support containerEnv format
    if "containerEnv" in data and isinstance(data["containerEnv"], dict):
        env_data = data["containerEnv"].copy()
    else:
        log(
            "ERR",
            f"Invalid JSON format in {json_file}. The file must contain a 'containerEnv' object.",
        )
        import sys

        sys.exit(1)

    # Include cli_version if it exists at the top level
    if "cli_version" in data:
        env_data["CLI_VERSION"] = data["cli_version"]

    lines = generate_exports(env_data, export_prefix=not no_export)

    # Ask for confirmation before writing to file
    if os.path.exists(output_file):
        if not confirm_action(f"This will overwrite the existing file at:\n{output_file}"):
            import sys

            sys.exit(1)
    else:
        if not confirm_action(f"This will create a new file at:\n{output_file}"):
            import sys

            sys.exit(1)

    try:
        with open(output_file, "w") as f:
            f.write("\n".join(lines) + "\n")
        log("OK", f"Wrote {len(lines)} exports to {output_file}")
    except Exception as e:
        log("ERR", f"Failed to write to {output_file}: {e}")
        import sys

        sys.exit(1)


def find_project_root(path: str) -> str:
    """Find the project root directory."""
    # If path is not provided, use current directory
    if not path:
        path = os.getcwd()

    # If path is a file, use its directory
    if os.path.isfile(path):
        path = os.path.dirname(path)

    # Check if the path has a .devcontainer directory
    if os.path.isdir(os.path.join(path, ".devcontainer")):
        return path

    log("ERR", f"Could not find a valid project root at {path}")
    log("INFO", "A valid project root must contain a .devcontainer directory")
    import sys

    sys.exit(1)
