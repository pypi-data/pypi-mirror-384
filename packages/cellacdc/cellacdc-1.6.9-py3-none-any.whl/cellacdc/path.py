from collections import defaultdict
import os
import sys
import shutil

import subprocess

from natsort import natsorted

from . import is_mac, is_linux
from . import printl
from . import myutils

def listdir(path):
    return natsorted([
        f for f in os.listdir(path)
        if not f.startswith('.')
        and not f == 'desktop.ini'
        and not f == 'recovery'
    ])

def newfilepath(file_path, appended_text: str=None):
    if appended_text is None:
        appended_text=''
    
    if not os.path.exists(file_path):
        return file_path, appended_text
    
    folder_path = os.path.dirname(file_path)
    filename = os.path.basename(file_path)
    filename, ext = os.path.splitext(filename)

    if appended_text:
        if appended_text.startswith('_'):
            appended_text = appended_text.lstrip('_')

    if appended_text:
        new_filename = f'{filename}_{appended_text}{ext}'
        new_filepath = os.path.join(folder_path, new_filename)
        if not os.path.exists(new_filepath):
            return new_filepath, appended_text
    
    i = 0
    while True:
        if appended_text:
            new_filename = f'{filename}_{appended_text}_{i+1}{ext}'
        else:
            new_filename = f'{filename}_{i+1}{ext}'
        new_filepath = os.path.join(folder_path, new_filename)
        if not os.path.exists(new_filepath):
            return new_filepath, f'{appended_text}_{i+1}'
        i += 1

def show_in_file_manager(path):
    if is_mac:
        args = ['open', fr'{path}']
    elif is_linux:
        args = ['xdg-open', fr'{path}']
    else:
        if os.path.isfile(path):
            args = ['explorer', '/select,', os.path.realpath(path)]
        else:
            args = ['explorer', os.path.realpath(path)]
    subprocess.run(args)

def copy_or_move_tree(
        src: os.PathLike, dst: os.PathLike, copy=False,
        sigInitPbar=None, sigUpdatePbar=None
    ):
    if sigInitPbar is not None:
        sigInitPbar.emit(0)
    
    files_failed_move = {}
    files_info = {}
    for root, dirs, files in os.walk(src):
        for file in files:
            rel_path = os.path.relpath(root, src).replace('\\', '/')
            src_filepath = os.path.join(root, file)
            dst_filepath = os.path.join(dst, *rel_path.split('/'), file)
            files_info[src_filepath] = dst_filepath
    
    if sigInitPbar is not None:
        sigInitPbar.emit(len(files_info))
    for src_filepath, dst_filepath in files_info.items():
        os.makedirs(os.path.dirname(dst_filepath), exist_ok=True)
        if copy:
            shutil.copyfile(src_filepath, dst_filepath)
        else:
            try:
                os.rename(src_filepath, dst_filepath)
            except Exception as e:
                shutil.copyfile(src_filepath, dst_filepath)
                files_failed_move[src_filepath] = dst_filepath
        if sigUpdatePbar is not None:
            sigUpdatePbar.emit(1)
    return files_failed_move

def get_posfolderpaths_walk(folderpath):
    pos_folderpaths = defaultdict(set)
    for root, dirs, files in os.walk(folderpath):
        if not root.endswith('Images'):
            continue
        
        pos_folderpath = os.path.dirname(root)
        if not myutils.is_pos_folderpath(pos_folderpath):
            continue
        
        exp_path = os.path.dirname(pos_folderpath).replace('\\', '/')
        pos_foldername = os.path.basename(pos_folderpath)
        pos_folderpaths[exp_path].add(pos_foldername)
    
    for exp_path in pos_folderpaths.keys():
        pos_folderpaths[exp_path] = natsorted(pos_folderpaths[exp_path])
    
    return pos_folderpaths