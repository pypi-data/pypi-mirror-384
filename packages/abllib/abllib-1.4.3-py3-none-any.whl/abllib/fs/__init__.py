"""A module containing file system-related functionality"""

from abllib.fs.filename import sanitize
from abllib.fs.path import absolute

__exports__ = [
    absolute,
    sanitize
]
