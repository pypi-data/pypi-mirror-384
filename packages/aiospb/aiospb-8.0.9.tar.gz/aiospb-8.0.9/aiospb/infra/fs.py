"""Implementation of ports with file system (files and directories)"""
import logging
import os
import os.path

import aiofiles

from aiospb.nodes.ports import Historian

from ..shared import DataType, DeviceKey, Metric, Quality

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)


class FSHistorian(Historian):
    """Store historical changes in plain files from a operating system"""

    def __init__(self, directory: str):
        self._dir = directory

    def _get_fn(self, key: DeviceKey) -> str:
        return os.path.join(self._dir, key.path.replace("/", "--") + ".csv")

    def _create_line(self, metric: Metric) -> str:
        quality = (
            metric.properties["quality"].value if "quality" in metric.properties else ""
        )
        return f"{metric.timestamp};{metric.value};{metric.data_type.name};{quality};{metric.alias};{metric.name}\n"

    async def load(self, key: DeviceKey) -> list[Metric]:
        metrics = []
        try:
            async with aiofiles.open(self._get_fn(key), "r") as f:
                while True:
                    line = await f.readline()

                    try:
                        values = line[:-1].split(";")

                        if len(values) == 1:
                            break

                        ts, value, datatype, quality, alias, name = values
                        m = Metric(
                            int(ts),
                            DataType[datatype].convert_value(value),
                            DataType[datatype],
                            int(alias),
                            name,
                            is_historical=True,
                        )
                        if quality:
                            m.quality = Quality(int(quality))
                        metrics.append(m)
                    except Exception as e:
                        logger.warning(f'Lost line "{line}" from history')
                        logger.exception(e)
        except FileNotFoundError:
            ...

        return metrics

    async def save(self, key: DeviceKey, metrics: list[Metric]):
        fn = os.path.join(self._dir, key.path.replace("/", "--") + ".csv")
        async with aiofiles.open(fn, "a") as f:
            lines = [self._create_line(metric) for metric in metrics]
            await f.writelines(lines)

    async def clear(self, key: DeviceKey):
        os.remove(self._get_fn(key))

    def get_device_keys(self, node_key: DeviceKey) -> list[DeviceKey]:
        keys = []
        for fn in os.listdir(self._dir):
            if "." in fn:
                key = fn[: fn.index(".")]
                key = key.replace("--", "/")
                device_key = DeviceKey(key)
                if device_key.node == node_key:
                    keys.append(device_key)
        return keys
