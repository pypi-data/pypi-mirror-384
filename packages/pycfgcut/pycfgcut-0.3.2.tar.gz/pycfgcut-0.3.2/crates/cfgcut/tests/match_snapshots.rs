use std::fs;
use std::path::PathBuf;

use assert_cmd::Command;
use insta::assert_snapshot;
use serde::Deserialize;

fn workspace_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("crate dir has parent")
        .parent()
        .expect("workspace root exists")
        .to_path_buf()
}

fn cfgcut_cmd() -> Command {
    let mut cmd = Command::cargo_bin("cfgcut").expect("binary built");
    cmd.current_dir(env!("CARGO_MANIFEST_DIR"));
    cmd
}

#[derive(Debug, Deserialize)]
struct Manifest {
    #[serde(rename = "case")]
    cases: Vec<Case>,
}

#[derive(Debug, Deserialize)]
struct Case {
    name: String,
    matches: Vec<String>,
    input: String,
}

fn load_manifest() -> Manifest {
    let path = workspace_root().join("tests/fixtures/synthetic_manifest.toml");
    let raw = fs::read_to_string(&path)
        .unwrap_or_else(|err| panic!("failed to read {}: {err}", path.display()));
    toml::from_str(&raw).unwrap_or_else(|err| panic!("failed to parse {}: {err}", path.display()))
}

#[test]
fn synthetic_snapshots() {
    let manifest = load_manifest();
    let root = workspace_root();

    for case in manifest.cases {
        let mut cmd = cfgcut_cmd();
        for pattern in &case.matches {
            cmd.args(["-m", pattern]);
        }
        cmd.arg(root.join(&case.input));

        let output = cmd.assert().success().get_output().stdout.clone();
        let stdout = String::from_utf8_lossy(&output);

        let mut snapshot = format!("input: {}\n", case.input);
        for pattern in &case.matches {
            snapshot.push_str("match: ");
            snapshot.push_str(pattern);
            snapshot.push('\n');
        }
        snapshot.push_str("---\n");
        snapshot.push_str(stdout.trim_end_matches('\n'));
        snapshot.push('\n');

        assert_snapshot!(case.name, snapshot);
    }
}
