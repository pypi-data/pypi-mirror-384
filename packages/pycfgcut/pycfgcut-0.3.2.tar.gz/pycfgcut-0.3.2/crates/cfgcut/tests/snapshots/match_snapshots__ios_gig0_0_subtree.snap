---
source: crates/cfgcut/tests/match_snapshots.rs
expression: snapshot
---
input: tests/fixtures/cisco_ios/full_lab.conf
match: interface GigabitEthernet0/0|>>|
---
! cfgcut matches for full_lab.conf
interface GigabitEthernet0/0
 description core-uplink
 ip address 10.0.0.2 255.255.255.252
 no shutdown
